### PR TITLE
[BEAM-13072][Playground] executor builder

### PR DESCRIPTION
**Pull request by daria-malkova**
[BEAM-13072] Improving executors architecture by implementing Builder pattern for flexible creation of executor object.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/go.mod [hunk 1]
```diff
@@ -18,6 +18,15 @@ module beam.apache.org/playground/backend
 go 1.16
 
 require (
+	github.com/google/uuid v1.3.0
+	github.com/kr/pretty v0.1.0 // indirect
+	github.com/tkanos/gonfig v0.0.0-20210106201359-53e13348de2f
+	golang.org/x/net v0.0.0-20210405180319-a5a99cb37ef4 // indirect
```
Comment (ilya-kozyrev) [inline playground/backend/go.mod:24]: Should we add indirect dependencies here?

File: playground/backend/internal/environment/beam.go [hunk 2]
```diff
@@ -15,14 +15,66 @@
...
+import (
+	pb "beam.apache.org/playground/backend/internal/api"
+	"github.com/tkanos/gonfig"
+	"os"
+	"strings"
+)
+
+// CmdConfig contains needed for compiling and execution of the code commands
+type CmdConfig struct {
+	CompileCmd  string   `json:"compile_cmd"`
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/beam.go:27]: let's use the full name in JSON representation too.

File: playground/backend/internal/executors/executor_builder.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+//withClassName adds file name to executor
+func (b *RunBuilder) withClassName(name string) *RunBuilder {
+	b.actions = append(b.actions, func(e *Executor) {
+		e.runArgs.fileName = name
+	})
+	return b
+}
+
+//withClassName adds file name to executor
+func (b *RunBuilder) withGraphOutput() *RunBuilder {
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/executors/executor_builder.go:106]: it seems that this comment is for another method.

File: playground/backend/internal/environment/beam.go [hunk 4]
```diff
@@ -15,14 +15,66 @@
...
 }
 
 // NewBeamEnvs is a BeamEnvs constructor
 func NewBeamEnvs(apacheBeamSdk pb.Sdk) *BeamEnvs {
-	return &BeamEnvs{ApacheBeamSdk: apacheBeamSdk}
+	config := readConfig(apacheBeamSdk)
+	return &BeamEnvs{ApacheBeamSdk: apacheBeamSdk, CmdConfig: config}
+}
+
+func readConfig(apacheBeamSdk pb.Sdk) *CmdConfig {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/beam.go:45]: I mentioned before, that all functionality to extract envs from os envs should be implemented in environment service.

File: playground/backend/internal/executors/executor_builder.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+//withWorkingDir adds dir path to executor
+func (b *RunBuilder) withWorkingDir(dir string) *RunBuilder {
+	b.actions = append(b.actions, func(e *Executor) {
+		e.runArgs.dirPath = dir
+	})
+	return b
+}
+
+//withClassName adds file name to executor
+func (b *ValidatorBuilder) withSdkValidators(validators *[]validators.Validator) *ValidatorBuilder {
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/executors/executor_builder.go:122]: it seems that this comment is for another method.

File: playground/backend/internal/environment/beam.go (hunk 4 above)
Comment (ilya-kozyrev) [reply playground/backend/internal/environment/beam.go:45]: this structure is DTO

File: playground/backend/internal/environment/beam.go [hunk 6]
```diff
@@ -15,14 +15,66 @@
...
+		cmdConfig.RunArgs[1] += jars
+	case pb.Sdk_SDK_GO:
+	case pb.Sdk_SDK_PYTHON:
+	case pb.Sdk_SDK_SCIO:
+	}
+
+	return &cmdConfig
+}
+
+func getCmdConfig(apacheBeamSdk pb.Sdk) CmdConfig {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/beam.go:65]: the same

File: playground/backend/internal/executors/executor.go [hunk 7]
```diff
@@ -17,81 +17,52 @@
...
 	"os/exec"
 )
 
-type validatorWithArgs struct {
-	validator func(filePath string, args ...interface{}) error
-	args      []interface{}
+//cmdSetting for base cmd code execution
+type cmdSetting struct {
+	fileName    string
+	dirPath     string
```
Comment (ilya-kozyrev) [inline playground/backend/internal/executors/executor.go:27]: is dir path is working dir? let's name it as a workingDir then

File: playground/backend/internal/executors/executor.go [hunk 8]
```diff
@@ -17,81 +17,52 @@
...
-	"fmt"
+	"beam.apache.org/playground/backend/internal/validators"
 	"os/exec"
 )
 
-type validatorWithArgs struct {
-	validator func(filePath string, args ...interface{}) error
-	args      []interface{}
+//cmdSetting for base cmd code execution
+type cmdSetting struct {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/executors/executor.go:25]: It seems like it's not setting. May be plural? Or CmdConfiguration?

File: playground/backend/internal/executors/executor_builder.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	b.actions = append(b.actions, func(e *Executor) {
+		e.runArgs.fileName = name
+	})
+	return b
+}
+
+//withGraphOutput adds the need of graph output to executor
+func (b *RunBuilder) withGraphOutput() *RunBuilder {
+	b.actions = append(b.actions, func(e *Executor) {
+		//smth
```
Comment (KhaninArtur) [inline playground/backend/internal/executors/executor_builder.go:108]: Let's replace `smth` with `TBD`

Comment (ilya-kozyrev) [reply playground/backend/internal/executors/executor_builder.go:108]: TODO

File: playground/backend/internal/environment/beam.go [hunk 10]
```diff
@@ -15,14 +15,26 @@
...
 // BeamEnvs contains all environments related of ApacheBeam. These will use to run pipelines
 type BeamEnvs struct {
 	ApacheBeamSdk pb.Sdk
+	CmdConfig     *CmdConfig
 }
 
 // NewBeamEnvs is a BeamEnvs constructor
 func NewBeamEnvs(apacheBeamSdk pb.Sdk) *BeamEnvs {
-	return &BeamEnvs{ApacheBeamSdk: apacheBeamSdk}
+	config := readConfig(apacheBeamSdk)
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/beam.go:38]: wrong place, keep this object from arguments

File: playground/backend/internal/environment/environment_service.go [hunk 11]
```diff
@@ -92,3 +94,40 @@ func getSdkEnvsFromOsEnvs() *BeamEnvs {
 
 	return NewBeamEnvs(sdk)
 }
+
+func readConfig(apacheBeamSdk pb.Sdk) *CmdConfig {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/environment_service.go:98]: rename it

File: playground/backend/internal/environment/environment_service.go [hunk 12]
```diff
@@ -92,3 +94,40 @@ func getSdkEnvsFromOsEnvs() *BeamEnvs {
...
+	cmdConfig := getCmdConfig(apacheBeamSdk)
+	switch apacheBeamSdk {
+	case pb.Sdk_SDK_JAVA:
+		cmdConfig.CompileArgs = append(cmdConfig.CompileArgs,
+			getEnv("BEAM_SDK", "/opt/apache/beam/jars/beam-sdks-java-harness.jar"))
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/environment_service.go:103]: keys and default values to constants

File: playground/backend/internal/environment/environment_service.go [hunk 13]
```diff
@@ -92,3 +94,40 @@ func getSdkEnvsFromOsEnvs() *BeamEnvs {
...
+
+	return &cmdConfig
+}
+
+func getCmdConfig(apacheBeamSdk pb.Sdk) CmdConfig {
+	CmdConfigPath := getEnv("CONFIG_FOLDER", "./") + apacheBeamSdk.String() + ".json"
+	cmdConfig := CmdConfig{}
+	err := gonfig.GetConf(CmdConfigPath, &cmdConfig)
+	if err != nil {
+		panic(err)
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/environment_service.go:123]: don't use panic

File: playground/backend/internal/environment/environment_service.go [hunk 14]
```diff
@@ -92,3 +94,40 @@ func getSdkEnvsFromOsEnvs() *BeamEnvs {
...
+	case pb.Sdk_SDK_GO:
+	case pb.Sdk_SDK_PYTHON:
+	case pb.Sdk_SDK_SCIO:
+	}
+
+	return &cmdConfig
+}
+
+func getCmdConfig(apacheBeamSdk pb.Sdk) CmdConfig {
+	CmdConfigPath := getEnv("CONFIG_FOLDER", "./") + apacheBeamSdk.String() + ".json"
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/environment_service.go:119]: key and default value to const

Comment (ilya-kozyrev) [inline playground/backend/internal/environment/environment_service.go:119]: don't hardcode .json

File: playground/backend/internal/environment/environment_service.go [hunk 15]
```diff
@@ -92,3 +94,40 @@ func getSdkEnvsFromOsEnvs() *BeamEnvs {
...
+	cmdConfig := CmdConfig{}
+	err := gonfig.GetConf(CmdConfigPath, &cmdConfig)
+	if err != nil {
+		panic(err)
+	}
+	return cmdConfig
+}
+
+func getEnv(key, fallback string) string {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/environment_service.go:128]: don't fully understand why we need this method, however if it reasonable - let's use this to read all envs

**Pull request closed without merging**